### PR TITLE
Rewrite

### PR DESCRIPTION
**Pull request by idubinskiy**

### CODE REVIEW

File: channel.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+// Channel represents a subscription to a Pusher channel.
+type Channel interface {
+	// Subscribe attempts to subscribe to the channel if the subscription is not
+	// already active. Authentication will be attempted for private and presence
+	// channels. Note that a nil error does not mean that the subscription was
+	// succesful, just that the subscription request was sent.
+	Subscribe() error
+	// Unsubscribe attempts to unsubscribe from the channel. Note that a nil error
+	// does not mean that the unsubscription was succesful, just that the request
```
Comment (melindabernrdo) [inline channel.go]: *successful

File: channel.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+	"strings"
+	"sync"
+)
+
+// Channel represents a subscription to a Pusher channel.
+type Channel interface {
+	// Subscribe attempts to subscribe to the channel if the subscription is not
+	// already active. Authentication will be attempted for private and presence
+	// channels. Note that a nil error does not mean that the subscription was
+	// succesful, just that the subscription request was sent.
```
Comment (melindabernrdo) [inline channel.go]: *successful

File: channel.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	// does not mean that the unsubscription was succesful, just that the request
+	// was sent.
+	Unsubscribe() error
+	// Bind returns a channel to which all the data from all matching events received
+	// on the channel will be sent.
+	Bind(event string) chan json.RawMessage
+	// Unbind removes bindings for an event. If chans are passed, only those bindings
+	// will be removed. Otherwise, all bindings for an event will be removed.
+	Unbind(event string, chans ...chan json.RawMessage)
+	// SendEvent sends an event to the channel.
```
Comment (melindabernrdo) [inline channel.go]: Was SendEvent renamed to Trigger?

Comment (melindabernrdo) [inline channel.go]: was SendEvent comment intended for Trigger?

File: channel.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+	Trigger(event string, data interface{}) error
```
Comment (melindabernrdo) [inline channel.go]: Also, thoughts on adding comments to the Trigger and HandleEvent functions to be consistent with the comments/annotations for all the other function definitions above?

File: channel.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+func (c *channel) Trigger(event string, data interface{}) error {
+	return c.client.SendEvent(event, data, c.name)
+}
+
+type privateChannel struct {
+	*channel
+}
+
+func (c *privateChannel) Subscribe() error {
```
Comment (melindabernrdo) [inline channel.go:130]: I may be missing something, but is there a place to add an authorization header with the auth request?

File: channel.go (hunk 1 above)
Comment (idubinskiy) [reply channel.go]: I hate that word.

File: channel.go (hunk 5 above)
Comment (idubinskiy) [reply channel.go:130]: There is not. It's a "todo" in Features in the README. I mistakenly assumed that the Objective-C and Java APIs didn't have this, so it was a nice-to-have. But it does seem like it's pretty necessary. I'll figure out a good way to add it.

File: channel.go (hunk 4 above)
Comment (idubinskiy) [reply channel.go]: For `Trigger`, that's the `SendEvent` comment. As far as `handleEvent`, that won't show up in the GoDoc, so ¯\\\_(ツ)\_/¯

File: channel.go (hunk 3 above)
Comment (idubinskiy) [reply channel.go]: Yup. Copy-pasta error.

**Pull request merged**